jetty-server/src/main/java/org/eclipse/jetty/server/HttpChannelState.java

```diff
@@ -576,6 +578,47 @@ public void dispatch(ServletContext context, String path)
             scheduleDispatch();
     }
 
+    public boolean asyncErrorDispatch(String path)
+    {
+        boolean dispatch = false;
+        AsyncContextEvent event;
+        try (Locker.Lock lock = _locker.lock())
+        {
+            if (LOG.isDebugEnabled())
+                LOG.debug("errorDispatch {} -> {}", toStringLocked(), path);
+
+            if (_async != Async.STARTED)
+                return false;
+
+            event = _event;
+            _async = Async.DISPATCH;
+
+            if (path != null)
+                _event.setDispatchPath(path);
+
+            _ignoreComplete = true;
+            switch (_state)
+            {
+                case DISPATCHED:
+                case ASYNC_IO:
+                case ASYNC_WOKEN:
+                    break;
+                case ASYNC_WAIT:
+                    _state = State.ASYNC_WOKEN;
+                    dispatch = true;
+                    break;
+                default:
+                    LOG.warn("asyncErrorDispatch when complete {}", this);
+                    break;
+            }
+        }
+
+        cancelTimeout(event);
+        if (dispatch)
+            scheduleDispatch();
+        return true;
+    }
```

Comment:
I would have expected an additional state here.
This method is only ever called from `doError()` and it does a dispatch. Is the `DispatcherType` (`ERROR` at this point) preserved?
Is there any risk that we end up in a loop where we end up calling this method again and again?

Reply:
Error page loop detection is already implemented.  We can only loop if the error page is different every time... so it is possible that errorPageA does an error to errorPageB which does an error to errorPageA etc. etc.

checking the dispatcherType....